modules/core/src/main/java/org/apache/synapse/message/processor/impl/forwarder/ForwardingService.java

```diff
@@ -609,13 +533,190 @@ public void dispatch(MessageContext messageContext) {
 			 * this by implementing a target inferring
 			 * mechanism.
 			 */
-
 			log.error("Neither targetEndpoint defined in the MessageProcessor configuration nor " +
 			          "Property " + ForwardingProcessorConstants.TARGET_ENDPOINT +
 			          " is found in the message context, hence deactivating the MessageProcessor");
 			deactivateMessageProcessor(messageContext);
 		}
-		return;
+	}
+
+	/**
+	 * Try to dispatch message to the given endpoint
+	 *
+	 * @param messageToDispatch MessageContext containing message to forward
+	 * @param endpoint                endpoint to forward message to
+	 */
+	private void tryToDispatchToEndpoint(MessageContext messageToDispatch, Endpoint endpoint) {
+
+		isSuccessful = false;
+		MessageContext outCtx = null;
+
+		try {
+			// For each retry we need to have a fresh copy of the original message
+			messageToDispatch.setEnvelope(MessageHelper.cloneSOAPEnvelope(messageToDispatch.getEnvelope()));
+			setSoapHeaderBlock(messageToDispatch);
+			updateAxis2MessageContext(messageToDispatch);
+
+			if (messageConsumer != null && messageConsumer.isAlive()) {
+				outCtx = sender.send(endpoint, messageToDispatch);
+			}
+
+			/*
+			 * Validation as message is forwarded successfully depends on
+			 * 1. is message forwarding validation is required
+			 * 2. is there a outCtx
+			 * 3. has some error happened inside blocking sender
+			 * 4. HTTP_SC code is in success family or configured to consider as a success
+			 * 5. Some exception has happened during message forwarding
+			 */
+
+			//For Protocols like JMS etc no need of validating response
+			if (isResponseValidationNotRequired) {
+				isSuccessful = true;
+				onForwardSuccess(endpoint);
+			}
+
+			//there is no response
+			if (!isResponseValidationNotRequired && outCtx == null) {
+				isSuccessful = true;
+				onForwardSuccess(endpoint);
+			}
+
+			//there is a response (In message context) but failed to send with no exception thrown
+			if (!isResponseValidationNotRequired && outCtx != null
+					&& "true".equals(outCtx.getProperty(SynapseConstants.BLOCKING_SENDER_ERROR))) {
+				log.error("Blocking Sender Error " + outCtx.getProperty(SynapseConstants.ERROR_EXCEPTION));
+				isSuccessful = false;
+				onForwardFailure();
+				//invoke fault sequence of MP
+				sendThroughFaultSeq(outCtx);
+				return;
+			}
+
+			//there is a response so check on status codes
+			if (!isResponseValidationNotRequired && outCtx != null && validateResponse(outCtx)) {
+				isSuccessful = true;
+				sendThroughReplySeq(outCtx);
+				onForwardSuccess(endpoint);
+			}
+
```

Comment:
We can refactor above multiple if blocks like this for better performance
```
if (isResponseValidationNotRequired) {
    isSuccessful = true;
    onForwardSuccess(endpoint);
} else {
    if (outCtx == null) {
       //do something
    } else {
       
    }
}
```

Reply:
This is intentionally done. If we implement with if else ladder logic is harder to understand (readability). 
			/*
			 * Validation as message is forwarded successfully depends on
			 * 1. is message forwarding validation is required
			 * 2. is there a outCtx
			 * 3. has some error happened inside blocking sender
			 * 4. HTTP_SC code is in success family or configured to consider as a success
			 * 5. Some exception has happened during message forwarding
			 */

Reply:
is this tryToDispatchToEndpoint in mediation runtime flow?
If so are we have made the trade-off of readability over performance?